src/bartiq/transform.py

```diff
@@ -0,0 +1,160 @@
+        subroutine (Dict[str, Any]): The subroutine to which the resources will be added.
+
+    Returns:
+        Dict[str, Any]: The subroutine with aggregated resources.
+    """
+    if "resources" not in subroutine:
+        return subroutine
+
+    # Initialize a dictionary for aggregated resources
+    aggregated_resources = {res["name"]: copy.deepcopy(res) for res in subroutine["resources"]}
+    # Iterate over the original resources in the subroutine
+    for resource in subroutine["resources"]:
+        resource_name = resource["name"]
+        resource_type = resource["type"]
+
+        try:
+            resource_expr = sympy.sympify(resource["value"])
+        except (sympy.SympifyError, TypeError) as e:
+            raise ValueError(f"Invalid resource value: {resource['value']}") from e
+
+        if resource_name in expanded_aggregation_dict:
+            mapping = expanded_aggregation_dict[resource_name]
+            for sub_res, multiplier in mapping.items():
+                if sub_res in aggregated_resources and resource_type == "additive":
+                    aggregated_resources[sub_res]["value"] = (
+                        sympy.sympify(aggregated_resources[sub_res]["value"]) + multiplier * resource_expr
+                    )
+                else:
+                    aggregated_resources[sub_res] = {
+                        "name": sub_res,
+                        "type": resource_type,
+                        "value": multiplier * resource_expr,
+                    }
+            del aggregated_resources[resource_name]
+
+    subroutine["resources"] = list(aggregated_resources.values())
+    return subroutine
+
+def _process_program(expanded_aggregation_dict: Dict[str, Dict[str, Any]], program: Dict[str, Any]) -> Dict[str, Any]:
+    program = _add_aggregated_resources_to_subroutine(expanded_aggregation_dict, program)
+    if "children" in program:
+        program["children"] = [_process_program(expanded_aggregation_dict, child) for child in program["children"]]
+
+    return program
+
+
+def _expand_aggregation_dict(aggregation_dict: Dict[str, Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
+    """
+    Expand the aggregation dictionary to handle nested resources.
+
+    Args:
+        aggregation_dict (Dict[str, Dict[str, Any]]): The input aggregation dictionary.
+
+    Returns:
+        Dict[str, Dict[str, Any]]: The expanded aggregation dictionary.
+    """
+    expanded_dict = {}
+    for resource in aggregation_dict.keys():
+        expanded_dict[resource] = _expand_resource(resource, aggregation_dict, set())
+    return expanded_dict
+
+def _expand_resource(resource: str, aggregation_dict: Dict[str, Dict[str, Any]], visited: set) -> Dict[str, Any]:
+    """
+    Recursively expand resource mapping to handle nested resources and detect circular dependencies.
+
+    Args:
+        resource (str): The resource to expand.
```

Comment:
[style] No need to provide type of the input param here, if you have have typing in the function signature.